latency/latency.go

```diff
@@ -0,0 +1,115 @@
+// Package latency implements protocol latency measurements
+package latency
+
+import (
+	"math"
+
+	"github.com/relab/hotstuff"
+	"github.com/relab/hotstuff/backend"
+	"github.com/relab/hotstuff/eventloop"
+	"github.com/relab/hotstuff/logging"
+	"github.com/relab/hotstuff/modules"
+)
+
+func init() {
+	modules.RegisterModule("latencymeasurement", NewLatencyMeasurement)
+}
+
+// NewLatencyMeasurement initializes the latency measurement.
+func NewLatencyMeasurement() modules.LatencyMeasurement {
+	return &latencyMeasurement{latencyStats: make(map[hotstuff.ID]map[hotstuff.ID]*statInfo)}
+}
+
+type latencyMeasurement struct {
+	configuration modules.Configuration
+	opts          *modules.Options
+	logger        logging.Logger
+	eventLoop     *eventloop.EventLoop
+	latencyStats  map[hotstuff.ID]map[hotstuff.ID]*statInfo
+	isInitDone    bool
+}
+
+type statInfo struct {
+	count    uint64
+	mean     float64
+	m2       float64
+	retValue uint64
+}
+
+func (l *latencyMeasurement) InitModule(mods *modules.Core) {
+	mods.Get(
+		&l.configuration,
+		&l.opts,
+		&l.logger,
+		&l.eventLoop,
+	)
+	l.eventLoop.RegisterHandler(hotstuff.LatencyVectorEvent{}, func(event any) {
+		l.handleNewMeasurement(event.(hotstuff.LatencyVectorEvent))
+	})
+	l.eventLoop.RegisterObserver(backend.ConnectedEvent{}, func(_ any) {
+		l.initialize()
+	})
+}
+
+func (l *latencyMeasurement) initialize() {
+	config := l.configuration.Replicas()
+	for id := range config {
+		tempStats := make(map[hotstuff.ID]*statInfo)
+		for id1 := range config {
+			tempStats[id1] = &statInfo{}
+		}
+		l.latencyStats[id] = tempStats
+	}
+	l.logger.Infof("stat are %v\n ", l.latencyStats)
+	l.isInitDone = true
+}
+
+func (l *latencyMeasurement) handleNewMeasurement(event hotstuff.LatencyVectorEvent) {
+	if !l.opts.IsLatencyCalculationEnabled() || !l.isInitDone {
+		return
+	}
+	latencyVector := event.LatencyVector
+	for id, newValue := range latencyVector {
+		stats := l.latencyStats[event.Creator][hotstuff.ID(id)]
+		if stats == nil {
+			return
+		}
+		l.ComputeStats(stats, float64(newValue))
+	}
+}
+
+func (l *latencyMeasurement) GetLatencyMatrix() map[hotstuff.ID]map[hotstuff.ID]uint64 {
```

Comment:
I suggest to not expose the matrix as a `map[hotstuff.ID]map[hotstuff.ID]uint64` but instead define a new type that has access methods:

```go
type LatencyMatrix map[hotstuff.ID]map[hotstuff.ID]uint64

// Latency returns the latency between replicas a and b.
func (lm *LatencyMatrix) Latency(a, b hotstuff.ID) time.Duration {}

// LatencyVector returns the latency vector for replica a.
func (lm *LatencyMatrix) LatencyVector(a hotstuff.ID) []time.Duration {}
```

This way you avoid exposing the internal datastructure of the matrix, enabling future optimization without impacting the code using the matrix.